### PR TITLE
Fix 367

### LINKED ISSUE #367: 'ranking_confidence' is not in the result.pkl file after the update
I think after this line:
https://github.com/KosinskiLab/AlphaPulldown/blob/main/alphapulldown/folding_backend/alphafold_backend.py#L542
there is no 'ranking_confidence' in the result.pkl file, after recalculate_confidence

```
result_pkl.keys()
Out[4]: dict_keys(['distogram', 'experimentally_resolved', 'masked_msa', 'num_recycles', 'predicted_aligned_error', 'predicted_lddt', 'structure_module', 'plddt'])
```

### PR DESCRIPTION
**Pull request by dingquanyu**
1. moved the step of dumping pickles into postprocessing after all the metrics are calculated
2. added test on the contents inside the result pickle
3. fixed #367

### CODE REVIEW

**Comment by dingquanyu** [comment]
Unittests on testRun1 and testRun2 were successful. I believe it's safe to merge.

File: test/check_predict_structure.py [hunk 1]
```diff
@@ -76,6 +77,14 @@ def _runCommonTests(self, result):
...
         #Check if the directory contains five files starting from result and ending with .pkl
         self.assertEqual(len([f for f in os.listdir(os.path.join(self.output_dir, dirname)) if f.startswith("result") and f.endswith(".pkl")]), 5)
+        #Check if the result pickle dictionary contains all the keys 
+        example_pickle = [f for f in os.listdir(os.path.join(self.output_dir, dirname)) if f.startswith("result") and f.endswith(".pkl")][0]
+        example_pickle = pickle.load(open((os.path.join(self.output_dir, dirname, example_pickle)), 'rb'))
+        if multimer_mode:
+            required_keys = ['distogram', 'experimentally_resolved', 'masked_msa', 'predicted_aligned_error', 'predicted_lddt', 'structure_module', 'plddt', 'aligned_confidence_probs', 'max_predicted_aligned_error', 'seqs', 'iptm', 'ranking_confidence']
+        else:
+            required_keys = ['distogram', 'experimentally_resolved', 'masked_msa', 'predicted_aligned_error', 'predicted_lddt', 'structure_module', 'plddt', 'aligned_confidence_probs', 'max_predicted_aligned_error', 'seqs','ranking_confidence']
```
[the remark this answers is not in the document]
Comment (dingquanyu) [reply test/check_predict_structure.py:86]: At the beginning I thought ptm will be one of these but it turned out ptm never 

> you don't check for 'ptm' here?

oh that's true. ptm is reported by AlphaFold as well. Forgot to add this

**Comment by DimaMolod** [comment]
> 1. moved the step of dumping pickles into postprocessing after all the metrics are calculated
> 2. added test on the contents inside the result pickle
> 3. fixed ['ranking_confidence' is not in the result.pkl file after the update #367](https://github.com/KosinskiLab/AlphaPulldown/issues/367)

if you dump pickles after all predictions are made, the resume (for prediction) will not work because it reads the prediction results from pkls here
https://github.com/KosinskiLab/AlphaPulldown/blob/fix-367/alphapulldown/folding_backend/alphafold_backend.py#L302

**Comment by dingquanyu** [comment]
> > 1. moved the step of dumping pickles into postprocessing after all the metrics are calculated
> > 2. added test on the contents inside the result pickle
> > 3. fixed ['ranking_confidence' is not in the result.pkl file after the update #367](https://github.com/KosinskiLab/AlphaPulldown/issues/367)
> 
> if you dump pickles after all predictions are made, the resume (for prediction) will not work because it reads the prediction results from pkls here https://github.com/KosinskiLab/AlphaPulldown/blob/fix-367/alphapulldown/folding_backend/alphafold_backend.py#L302

That's true. Maybe dump pickle as well in the prediction step? Anyway the pickles are gonna be overwritten again in the postprocess step.

**Comment by DimaMolod** [comment]
yes, dumping pickles twice would be the easiest solution, I guess. Could you please make sure that testRun_5 works fine after the update?

**Comment by dingquanyu** [comment]
> yes, dumping pickles twice would be the easiest solution, I guess. Could you please make sure that testRun_5 works fine after the update?

OK! Tests are running now

**Comment by dingquanyu** [comment]
> yes, dumping pickles twice would be the easiest solution, I guess. Could you please make sure that testRun_5 works fine after the update?

Now test 1,2 and 5 are all successful. That was a good catch btw.  I think now it's safe to merge

**Pull request merged**